packages/aws-cdk-lib/aws-cloudwatch/lib/variable.ts

```diff
@@ -0,0 +1,182 @@
+
+export enum VariableInputType {
+
+  /**
+     * Freeform text input box
+     */
+  INPUT = 'input',
+
+  /**
+     * A dropdown of pre-defined values, or values filled in from a metric search query
+     */
+  RADIO = 'radio',
+
+  /**
+     * A set of pre-defined radio buttons, which can also be defined from a metric search query
+     */
+  SELECT = 'select',
+}
+
+export enum VariableType {
+
+  /**
+     * A property variable changes the values of all instances of a property in the list of widgets in the dashboard.
+     */
+  PROPERTY = 'property',
+
+  /**
+     * A pattern variable is one that changes a regex pattern across the dashboard JSON
+     */
+  PATTERN = 'pattern',
+}
+
+/**
+ * A single dashboard variable
+ */
+export interface IVariable {
+
+  /**
+     * Return the variable JSON for use in the dashboard
+     */
+  toJson(): any;
+}
+
+export interface IDashboardVariable {
+
+  /**
+   * Type of the variable
+   */
+  readonly type: VariableType;
+
+  /**
+   * The way the variable value is selected
+   */
+  readonly inputType: VariableInputType;
+
+  /**
+   * Pattern or property value to replace
+   */
+  readonly value: string;
+
+  /**
+   * Unique id
+   */
+  readonly id: string;
+
+  /**
+   * Optional label in the toolbar
+   */
+  readonly label?: string;
+
+  /**
+   * Optional default value
+   */
+  readonly defaultValue: any;
+
+  /**
+   * Whether the variable is visible
+   */
+  readonly visible?: boolean;
+}
+
+export abstract class DashboardVariable implements IVariable {
+  private readonly baseProps: IDashboardVariable;
+
+  protected constructor(props: IDashboardVariable) {
+    this.baseProps = props;
+  }
+
+  toJson(): any {
+    return {
+      [this.baseProps.type]: this.baseProps.value,
```

Comment:
Can you explain this line for me? I'm not sure what `value` is supposed to do and I don't see this as part of the [API](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/CloudWatch-Dashboard-Body-Structure.html#CloudWatch-Dashboard-Properties-Variables-Structure)

Reply:
This line will resolve to either:
```
{
property: 'BucketName'
}
```
 if value specified is 'BucketName'. In this case the inputs from property variable will result into changing of BucketName property in the Dashboard json.

if variable type is pattern, then the value field corresponds to the pattern which has to be replaced in the dashboard. Say we want to replace all occurrences of word 'Placeholder' in the dashboard with another value from the variable. Then it's possible to do if variable json has:

```
{
pattern: "Placeholder"
}
```

From the API doc: 
for the property variable, here the `value` field is the `region`
```
"variables": [
     {
        "type": "property", 
        "property": "region",
        "inputType": "input",
        "id": "region",
        "label": "Region",
        "defaultValue": "us-east-1",
        "visible": true
    } 
],
```

for the pattern variable, here the `value` field is the `us-east-1`
```
"variables": [
    {
        "type": "pattern",
        "pattern": "us-east-1",
        "inputType": "input",
        "id": "region",
        "label": "Region",
        "defaultValue": "us-east-1",
        "visible": true
    }
],
```



So overall, the `value` field controls what exactly will change in the dashboard: some string, some property, dimension name and so on

Reply:
Thanks for the detailed reply @Stacy-D! I really appreciate it.